quesma/processors/tcp_passthrough_mysql_processor.go

```diff
@@ -70,18 +78,146 @@ func (t *TcpPassthroughMySqlProcessor) Handle(metadata map[string]interface{}, m
 	// tcp_passthrough_mysql_processor <- tcp_backend_connector <- real MySQL server
 	//
 	// and forwards them back to the MySQL client.
+
+	eofCount := 0
+
 	for {
 		fullPacketBytes, err := frontend_connectors.ReadMySqlPacket(conn)
 		if err != nil {
+			if err != io.EOF {
+				logger.Warn().Err(err).Msgf("Error reading MySQL packet from the MySQL TCP backend")
+			} else {
+				logger.Debug().Msg("Finished reading MySQL packets from the MySQL TCP backend (EOF), returning control to the frontend")
+			}
+			break
+		}
+
+		if metadata["handshake_processed"] == nil {
+			metadata["handshake_processed"] = true
+			fullPacketBytes = maybeProcessHandshake(fullPacketBytes)
+
+			response = append(response, fullPacketBytes...)
+			logger.Debug().Msg("Finished reading MySQL packets from the MySQL TCP backend (handshake), returning control to the frontend")
 			break
 		}
 
+		logger.Debug().Msgf("Received packet from MySQL backend of length %d", len(fullPacketBytes))
+
 		response = append(response, fullPacketBytes...)
+
+		// Early exit - stop reading packets if we figure out that the server has sent the entire response:
+		// an OK or ERR response or a result set (with two EOF packets).
+
+		// https://dev.mysql.com/doc/dev/mysql-server/8.4.3/page_protocol_basic_ok_packet.html
+		// Quote:
+		// > These rules distinguish whether the packet represents OK or EOF:
+		// > OK: header = 0 and length of packet > 7
+		// > EOF: header = 0xfe and length of packet < 9
+		//
+		// Note: length of packet > 7 seems like a mistake in the documentation, it should be length of packet >= 7
+
+		// OK packet
+		const OK_PACKET = 0x00
+		if len(fullPacketBytes) >= 7+4 && fullPacketBytes[4] == OK_PACKET {
```

Comment:
I would suggest to extract these literals `7`, `4` as const with meaningful name